### PR TITLE
Avoid `TrainablePipe.finish_update` getting called twice during training

### PR DESCRIPTION
**Pull request by danieldk**
## Description

PR #12136 fixed an issue where the tok2vec pipe was updated before gradient were accumulated. However, it introduced a new bug that cause `finish_update` to be called twice when using the training loop. This causes a fairly large slowdown.

The `Language.update` method accepts the `sgd` argument for passing an optimizer. This argument has three possible values:

- `Optimizer`: use the given optimizer to finish pipe updates.
- `None`: use a default optimizer to finish pipe updates.
- `False`: do not finish pipe updates.

However, the latter option was not documented and not valid with the existing type of `sgd`. I assumed that this was a remnant of earlier spaCy versions and removed handling of `False`.

However, with that change, we are passing `None` to `Language.update`. As a result, we were calling `finish_update` in both `Language.update` and in the training loop after all subbatches are processed.

This change restores proper handling/use of `False`. Moreover, the role of `False` is now documented and added to the type to avoid future accidents.

### Types of change

Bug fix

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [x] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

File: website/docs/api/language.mdx [hunk 1]
```diff
@@ -323,15 +323,15 @@ and custom registered functions if needed. See the
...
+| Name            | Description                                                                                                                                                                                           |
+| --------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `examples`      | A batch of [`Example`](/api/example) objects to learn from. ~~Iterable[Example]~~                                                                                                                     |
+| _keyword-only_  |                                                                                                                                                                                                       |
+| `drop`          | The dropout rate. ~~float~~                                                                                                                                                                           |
```
[the remark this answers is not in the document]
Comment (danieldk) [reply website/docs/api/language.mdx]: Done.

**Review by shadeMe — APPROVED** [review]

**Review by svlandeg — APPROVED** [review]
LGTM!
